Metrics Monitoring: A minimal viable product

simon-mo opened the pull request:
This PR sets up the basic structure of monitoring and provide basic functionalities:
- It exports query frontend metrics through a coupled node exporter. 
- It exports model container metrics (count, 3 types of latencies in seconds) through a node exporter in the same container but different python process.
- It initiates a mechanism to spin up a prometheus monitor and update it when necessary. 

RPC image needs to be rebuilt. A new frontend-exporter image is also created. 

Full implementation available in commit messages. 

[Simon Mo]

simon-mo committed:
[Metrics] A working node exporter done.

User can now just run clipper like regular run and spin up a
metric monitor under the condition:
1. There is a clipper exporter docker image.

<Simon Mo>

simon-mo committed:
[Metrics] A working node exporter done.

User can now just run clipper like regular run and spin up a
metric monitor under the condition:
1. There is a clipper exporter docker image.

<Simon Mo>

simon-mo committed:
[Metric] Model Exporter Done.

`rpc.py` is modified so that the metric monitor is put in. Metric
exporter now run along side with the model closure and communicate
via a pipe. The prometheus client is updated whenever a new replica
is added.

`docker_metric_utils.py` is the file for all metric related code that
is put in `docker_container_manager.py` to reduce clutter. Prometheus
listen to port 1390, a port specified in `container_manager.py` as a
constant.

The requirement for `prometheus_client` package is added to three
places: `requirements.txt`, `setup.py`, `RPCDockerfile`

I still need to work on integration test though.

<Simon Mo>

simon-mo committed:
[Metric] Model Exporter Done.

`rpc.py` is modified so that the metric monitor is put in. Metric
exporter now run along side with the model closure and communicate
via a pipe. The prometheus client is updated whenever a new replica
is added.

`docker_metric_utils.py` is the file for all metric related code that
is put in `docker_container_manager.py` to reduce clutter. Prometheus
listen to port 1390, a port specified in `container_manager.py` as a
constant.

The requirement for `prometheus_client` package is added to three
places: `requirements.txt`, `setup.py`, `RPCDockerfile`

I still need to work on integration test though.

<Simon Mo>

simon-mo committed:
[Metric] Add FrontendExporter Docker image

Changed the corresponding files for the frontend exporter to be built.
- `build_docker_image.sh` and `docker_metric_utils.py` agree on the name
'frontend-exporter'
- python file and Dockerfile are in place

<Simon Mo>

simon-mo committed:
[Metric] Add FrontendExporter Docker image

Changed the corresponding files for the frontend exporter to be built.
- `build_docker_image.sh` and `docker_metric_utils.py` agree on the name
'frontend-exporter'
- python file and Dockerfile are in place

<Simon Mo>

simon-mo committed:
[Metric] Add docstrings

simon-mo committed:
[Metric] Add docstrings

simon-mo committed:
[Metric] Add integration test for clipper metric

This is just a simple modification of basic_query example.
It tests for two replicas to see if the prometheus client
is monitoring all three node exporter.

simon-mo committed:
[Metric] Add integration test for clipper metric

This is just a simple modification of basic_query example.
It tests for two replicas to see if the prometheus client
is monitoring all three node exporter.

AmplabJenkins commented:
Can one of the admins verify this patch?

simon-mo committed:
[Metric] Format Code

simon-mo committed:
[Metric] Format Code

simon-mo committed:
[Metrics] Bug Fix, update the name accoridingly

simon-mo committed:
[Metrics] Bug Fix, update the name accoridingly

dcrankshaw commented:
jenkins okay to test

dcrankshaw commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -123,24 +125,38 @@ def start_clipper(self, query_frontend_image, mgmt_frontend_image,
+
+        # Metric Section
+        query_frontend_metric_name = "query_frontend_exporter-{}".format(
+            random.randint(0, 100000))
```
Let's use the same random integer as the `query_frontend_name`

simon-mo replied:
Fixed.

dcrankshaw commented on containers/python/front_end_exporter.py:
```diff
@@ -0,0 +1,69 @@
+import requests
```
The containers directory is for stuff related to model containers. Can you create a separate monitoring directory (`CLIPPER_ROOT/monitoring`) and put this file in it?

dcrankshaw replied:
Can you also put a short README in the monitoring directory that provides instructions on how to access the Prometheus server once it is up?

simon-mo replied:
Got it! This is a much better idea.

dcrankshaw commented on containers/python/rpc.py:
```diff
@@ -488,4 +503,37 @@ def start(self, model, host, port, model_name, model_version, input_type):
+    def collect(self):
+        curr = None
+        while self.pipe_conn.poll():
+            curr = self.pipe_conn.recv()
```
If there are multiple items in the pipe, won't we overwrite all except the last item? It seems like `curr` should be a list you append to?

simon-mo replied:
The purpose is to just get the the last item. Prometheus 'scrape' for the latest metric every 5 seconds and the metric is updated only for each `GET` call. 

Because the prediction time is variable, in the rpc loop, we push the most recent metric dictionary into the pipe after each prediction call. In the metric collector, we just need to get the most recent one. 

I changed the name to `lastest_metric_dict` for better clarity.

dcrankshaw commented on containers/python/rpc.py:
```diff
@@ -488,4 +503,37 @@ def start(self, model, host, port, model_name, model_version, input_type):
+        if curr:
+            for name, val in curr.items():
+                try:
+                    yield GaugeMetricFamily(name, 'help', value=val)
```
What is the "help" string for?

simon-mo replied:
This is a required argument for this constructor. But not available for end-user. 
I changed it to:
```python
yield GaugeMetricFamily(name=name, 
                                          documentation=name, # Required Argument
                                           value=val)
```
for clarity.

dcrankshaw reviewed (changes requested):
This looks like a great MVP for monitoring. I just had a few small comments.

simon-mo committed:
[Metric] Small Fixes

Update the code according to Dan's comments. Moved the front-end
container to a separate directory.

simon-mo committed:
[Metric] Small Fixes

Update the code according to Dan's comments. Moved the front-end
container to a separate directory.

simon-mo committed:
[Metric] Format Code

simon-mo committed:
[Metric] Format Code

simon-mo commented:
@dcrankshaw It looks like Jenkins isn't starting. Can you start the check manually?

dcrankshaw commented:
jenkins ok to test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/824/
Test FAILed.

simon-mo committed:
[Metric] Skipped Metrics to Let Jenkins Build Images

simon-mo committed:
[Metric] Skipped Metrics to Let Jenkins Build Images

simon-mo commented:
~~Don't merge. I commented out the metrics section to let Jenkins to build new docker images.~~ Resolved

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/825/
Test FAILed.

simon-mo committed:
[Metric] Revert the files; docker imgs built before test

simon-mo committed:
[Metric] Revert the files; docker imgs built before test

simon-mo committed:
Merge branch 'metrics' of https://github.com/simon-mo/clipper into metrics

simon-mo committed:
Move the comments; restart the tests

simon-mo committed:
Move the comments; restart the tests

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/826/
Test FAILed.

simon-mo committed:
[Metric]  Add version tag to the frontend-exporter

simon-mo committed:
[Metric]  Add version tag to the frontend-exporter

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/827/
Test PASSed.

simon-mo committed:
[Metric] Add clipper_metric to run_unittests.sh

simon-mo committed:
[Metric] Add clipper_metric to run_unittests.sh

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/828/
Test FAILed.

simon-mo committed:
[Metric] Trigger another CI check

Pretty sure it isn't the issue with my code. But it now appears
common enough.

simon-mo committed:
[Metric] Trigger another CI check

Pretty sure it isn't the issue with my code. But it now appears
common enough.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/829/
Test FAILed.

simon-mo commented:
@dcrankshaw Help please. 

It seems `clipper_admin_tests.py` causes the failure. It fails on the same error but with different unit tests that has nothing to do with the code I changed:
- test_inspect_instance_returns_json_dict: Bind for 0.0.0.0:37967 failed: port is already allocated. [829 log](https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/829/console).
- test_python_closure_deploys_successfully: Error starting userland proxy: listen tcp 0.0.0.0:38198: bind: address already in use. [828 log](https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/828/console).
- test_link_not_registered_model_to_app_fails : Error starting userland proxy: listen tcp 0.0.0.0:39526: bind: address already in use. [822 log (for batch predict)](https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/822/console).

dcrankshaw committed:
Merge branch 'develop' into metrics

dcrankshaw commented:
Yeah there's an issue with out Jenkins tests. It causes the master branch builder to be super flaky too. I haven't had time to chase it down. Can you file a issue with the above comment?

simon-mo commented:
Sure! Done.

dcrankshaw commented on containers/python/rpc.py:
```diff
@@ -515,13 +515,16 @@ def __init__(self, pipe_child_conn):
 
     def collect(self):
-        curr = None
+        lastest_metric_dict = None
```
s/lastest/latest

simon-mo replied:
Fixed

dcrankshaw commented on monitoring/README.md:
```diff
@@ -0,0 +1,13 @@
+This module is related Clipper's metric monitoring function. For full design, see [Design Doc.](https://docs.google.com/document/d/10whRxCc97gOJl4j2lY6R-v7cI_ZoAMVcNGPG_9oj6iY/edit?usp=sharing)
```
"is related to"

simon-mo replied:
fixed

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/830/
Test FAILed.

dcrankshaw commented on containers/python/rpc.py:
```diff
@@ -306,6 +312,15 @@ def run(self):
 
                         response.send(socket, self.event_history)
 
+                        pred_metric['model_pred_count'] += 1
```
You're accumulating the counts here, but just sending samples of the latencies. This is why I was confused about how you were dequeuing from the pipe. In order to handle the latencies correctly you should be creating histograms. For this MVP PR, let's just accumulate the count. So delete the `*_time` metrics.

simon-mo replied:
Got it. I was thinking about this recently. I deleted `*_time` metrics. 

If we can finalize (at least for now) all the metrics needed to be collected, I will initialize them in the prometheus collector process just once [1] and run update through each dictionary in the pipe. So that every metric record is updated in the metric process (especially matters for histogram and summary data type. 

[1]: the current process initialize a new `gauge` every time the metric is updated.

dcrankshaw replied:
Yeah that's definitely the right way to do it. Let's merge this PR, then you can start on one that tracks the timing metrics with histograms.

dcrankshaw commented on containers/python/rpc.py:
```diff
@@ -488,4 +503,40 @@ def start(self, model, host, port, model_name, model_version, input_type):
+                    yield GaugeMetricFamily(
+                        name=name,
+                        documentation=name,  # Required Argument
+                        value=val)
```
s/lastest/latest

simon-mo replied:
fixed

dcrankshaw reviewed (changes requested):
This is cool! I got it working and even just playing around with the bare bones Prometheus is pretty cool. I'm excited to start building out our monitoring infrastructure.

Just a couple minor comments and then this is good to go.

simon-mo committed:
Address comments, fix typo

simon-mo committed:
Merge branch 'metrics' of https://github.com/simon-mo/clipper into metrics

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/831/
Test FAILed.

simon-mo committed:
[Metric] change the redis port for integration-test

So that it won't clash with other testing progress

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/832/
Test PASSed.

dcrankshaw reviewed (approved):
LGTM. Fix the merge conflicts then I'll get this merged.

simon-mo committed:
Merge branch 'develop' into metrics

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/849/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/850/
Test PASSed.
